ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,403 @@
+            never_reset_tim[i] = true;
+            never_reset_pin_number(self->tim->pin->port, self->tim->pin->number);
+            break;
+        }
+    }
+}
+
+void common_hal_pulseio_pwmout_reset_ok(pulseio_pwmout_obj_t *self) {
+    //TODO: doesn't this need an equivalent release pin in microcontroller.c? 
+    //I don't see that implemented in any port. 
+    for(size_t i = 0 ; i < TIM_BANK_ARRAY_LEN; i++) {
+        if (mcu_tim_banks[i] == self->handle.Instance) {
+            never_reset_tim[i] = false;
+            break;
+        }
+    }
+}
+
+pwmout_result_t common_hal_pulseio_pwmout_construct(pulseio_pwmout_obj_t* self,
+                                                    const mcu_pin_obj_t* pin,
+                                                    uint16_t duty,
+                                                    uint32_t frequency,
+                                                    bool variable_frequency) {
+    if (frequency == 0 || frequency > 6000000) {
+        mp_raise_ValueError(translate("Invalid frequency supplied"));
+        //return PWMOUT_INVALID_FREQUENCY;
+    }
+
+    TIM_TypeDef * TIMx;
+    //TODO: style/safety question, use this instead?
+    //uint8_t tim_num_pins = TIM_PIN_ARRAY_LEN;
+    uint8_t tim_num = sizeof(mcu_tim_pin_list)/sizeof(*mcu_tim_pin_list);
+    //TODO: use an enum to make this prettier?
+    bool tim_chan_taken = false;
+    bool tim_taken_f_mismatch = false;
+    bool var_freq_mismatch = false;
+    bool first_time_setup = true;
+
+    for(uint i = 0; i < tim_num; i++) {
+        //if pin is same
+        if(mcu_tim_pin_list[i].pin == pin) {
+            //check if the timer has a channel active
+            if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] != 0) {
+                //is it the same channel? (or all channels reserved by a var-freq)
+                if(reserved_tim[mcu_tim_pin_list[i].tim_index-1] & 1<<(mcu_tim_pin_list[i].channel_index-1)) {
+                    tim_chan_taken = true;
+                    continue; //keep looking, might be another viable option
+                }
+                //If the frequencies are the same it's ok
+                if(tim_frequencies[mcu_tim_pin_list[i].tim_index-1] != frequency) {
+                    tim_taken_f_mismatch = true;
+                    continue; //keep looking
+                }
+                //you can't put a variable frequency on a partially reserved timer
+                if(variable_frequency) {
+                    var_freq_mismatch = true;
+                    continue; //keep looking
+                }
+                first_time_setup = false; //skip setting up the timer
+            }
+            //No problems taken, so set it up
+            self->tim = &mcu_tim_pin_list[i];
+            break;
+        }
+    }
+
+    //handle valid/invalid timer instance
+    //TODO: why doesn't PWM handle its own value errors like every other module?
+    //The samd value errors don't really fit for STM32.
```

Comment:
OK to change shared-bindings to fix this.

Reply:
This will also require changing both the NRF and atmel ports. I'll make an issue.